HTTP/3 with strict priorities (#2700)
quicwg/base-drafts: draft-ietf-quic-http.md

The PRIORITY (type=0x2) frame specifies the client-advised priority of a
request, server push or placeholder.

A PRIORITY frame identifies an element to prioritize and a placeholder upon
which it depends, or the root if no placeholder is specified.  A Prioritized
ID identifies a server push using a Push ID (see {{frame-push-promise}}),
or a placeholder using a Placeholder ID (see {{placeholders}}).

When a client initiates a request, a PRIORITY frame MAY be sent as the first
frame of the stream, creating a dependency on an existing element.  In order to
ensure that prioritization is processed in a consistent order, any subsequent

> PRIORITY frames for that request MUST be sent on the request stream.

Reviewer: The difficulty here is that you'll need to keep the request stream open (from the client-send side) for as long you want to reprioritize the request. This risks conflicting with the current text in section 5.1 

> An HTTP request/response exchange fully consumes a bidirectional QUIC
   stream.  After sending a request, a client MUST close the stream for
   sending.

That is not insurmountable, you just might need some extra text in the draft to deal with the ripple effects of the changes.

Author: I've now reverted this portion of the change, thanks for the suggestion.

Placeholder and server push streams can only be prioritized on the control
stream.

~~~~~~~~~~  drawing
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|P|D|W|P| Empty |         [Prioritized Element ID (i)]        ...
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                     [Placeholder ID (i)]                    ...
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|  [Weight (8)] | [Priority (8)]|
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
~~~~~~~~~~
{: #fig-priority title="PRIORITY frame payload"}

The PRIORITY frame payload has the following fields:

  P (Prioritized Element Type):
  : A one-bit field indicating the type of element being prioritized (see
    {{prioritized-element-types}}). When sent on a request stream, this MUST be
    set to `0`.

  D (Dependent On Placeholder):

>   : A one-bit field indicating whether a element depends upon a placeholder

Reviewer: ```suggestion
  : A one-bit field indicating whether an element depends upon a placeholder
```

    or the root of the tree (see {{element-dependency-types}}).

  W (Weight Indication):
  : A one-bit field indicating whether an 8-bit weight value is present.

  P (Priority Indication):
  : A one-bit field indicating whether an 8-bit strict priority is present.

  Empty:
  : A four-bit field which MUST be zero when sent and has no semantic value on
    receipt.

  Prioritized Element ID:
  : A variable-length integer that identifies the element being prioritized.
    The Push ID of a promised resource, a Placeholder ID of a placeholder,
    or is absent if sent on the request stream.

  Element Dependency ID:
  : A variable-length integer that identifies the Placeholder ID on which a
    dependency is being expressed, or is absent.  For details of dependencies,
    see {{priority}} and {{!RFC7540}}, Section 5.3.

  Weight:
  : An optional unsigned 8-bit integer representing a priority weight for the
    prioritized element (see {{!RFC7540}}, Section 5.3). Add one to the value
    to obtain a weight between 1 and 256.  When absent, indicates the resource

>     should be delivered all at once or not at all.

Reviewer 2: Server might not have all data in place (e.g., streaming via proxy).  To achieve "delivered all at once or not at all", it seems to me that proxy has to buffer all data, which is impossible if data is unlimited or too large.

Reviewer 3: Maybe a similar indication as in the HTTP/2 RFC makes sense:

"When absent, indicates the resource should be allocated the full available resources from its parent, as long as it is able to make progress."

Author: Thanks @rmarx suggestion taken.

of the associated Element ID fields.

| P Bit | Type Description | Prioritized Element ID Contents |
| ----- | ---------------- | ------------------------------- |
| 0     | Current stream   | Absent                          |
| 0     | Push stream      | Push ID                         |
| 1     | Placeholder      | Placeholder ID                  |
{: #prioritized-element-types title="Prioritized Element Types"}

| D  Bit | Type Description | Placeholder ID Contents |
| ------ | ---------------- | ----------------------- |
| 0      | Root of the tree | Absent                  |
| 1      | Placeholder      | Placeholder ID          |
{: #element-dependency-types title="Element Dependency Types"}

Note that unlike in {{!RFC7540}}, the root of the tree cannot be referenced
using 0, so the root of the tree cannot be reprioritized.

The PRIORITY frame can express relationships which might not be permitted based
on the stream on which it is sent or its position in the stream. These
situations MUST be treated as a connection error of type HTTP_MALFORMED_FRAME.
The following situations are examples of invalid PRIORITY frames:

> - A PRIORITY frame sent on a request stream with the Prioritized Element Type

Reviewer: Since the list is one bullet point long with this change, you could probably roll this back into prose

Author: I added back another one, but it's still a short list.  I believe one more will end up being added, so keeping the structure for now to minimize diffs.

## Prioritization {#priority}

HTTP/3 uses a priority scheme similar to that described in {{!RFC7540}}, Section
5.3, but replaces streams depending upon streams with strict priorities and
placeholders.  In the HTTP/3 priority scheme, a stream can be given a strict

> priority or can be designated as dependent upon a placeholder or the root.

Reviewer 4: ...or weights, apparently?

This information is expressed in the PRIORITY frame {{frame-priority}} which
identifies the element and the dependency. The elements that can be prioritized
are:

- Requests, identified by the stream of the PRIORITY frame
- Pushes, identified by the Push ID of the promised resource
  ({{frame-push-promise}})
- Placeholders, identified by a Placeholder ID

In HTTP/3, stream dependencies that were implicitly encoded by dependencies in
HTTP/2 are explicitly encoded with strict prioritization.  This simplifies
priority tree management, eliminates potential new race conditions introduced
by HTTP/3's multiple streams, and improves framing efficiency with more than

> 64 requests.

Reviewer 4: I'm not sure we need to be explicit about the 64 requests part.  Maybe just say "for long-lived connections"?

Reviewer 5: The can sentence can be removed: it's an editorial.

Author: Removed.

When a prioritized element is first created, it has a default initial weight

> of 16, priority of 16, and a default dependency. Requests and placeholders are

Reviewer 4: I'm surprised that they have an initial weight, when weight can be omitted.  Shouldn't the default be to have null weight?

Author: Agreed, I'll change to 0 weight, which is equivalent to omitting a weight.

dependent on the root of the priority tree; pushes are dependent on the client

> request on which the PUSH_PROMISE frame was sent.

Reviewer 4: ...except that your PR removes the ability to depend on streams, so this means push streams start out in an impossible-to-express state.  Not _necessarily_ a problem (we're doing it with the orphan placeholder), but make sure it's deliberate.

Author: Good point, will fix this.

Requests may override the default initial values by including a PRIORITY frame
(see {{frame-priority}}) at the beginning of the stream. These priorities
can be updated by sending a PRIORITY frame on the control stream.

Higher priority elements have all available data sent before elements of lower
priority.  When multiple elements have the same priority, if a weight is
specified, elements with weights are interleaved.  If a weight is not
specified, then elements are delivered sequentially and all available data
from one element is sent before any data from the next.  If a given priority
level has some elements with weights and some without, the two groups share

> the available bandwidth equally.

Reviewer 4: How weights and explicit priorities interact probably needs to come at the top of the section, since it's a pretty substantial departure from HTTP/2.

Reviewer 6: Would it make sense to change the requirement of the last sentence to something like: "If a given priority level has some elements with weights and some without, those without weights are sent before any of those with weights"?

I suggest this because it would be a simplification. Current text requiring the bandwidth to be evenly split between the two groups essentially suggests implementations to create one more WRR instance - I prefer avoiding that.

I think such a change is fine because I do not think most of the client would use the prioritization scheme in such a way, and because clients that indeed want in fact to split the bandwidth can use placeholders to accommodate that.

In HTTP/3, stream dependencies that were implicitly encoded by dependencies in
HTTP/2 are explicitly encoded with strict prioritization.

When a prioritized element is first created, it has a default initial weight
of 0, priority of 16, and a default dependency. Requests and placeholders are
dependent on the root of the priority tree; pushes are dependent on the same
parent as the client request on which the PUSH_PROMISE frame was sent and given
a priority value one smaller than the request stream.

Requests may override the default initial values by including a PRIORITY frame
(see {{frame-priority}}) at the beginning of the stream. These priorities
can be updated by sending a PRIORITY frame on the control stream.

Higher priority elements have all available data sent before elements of lower
priority.  When multiple elements have the same priority, if a weight is
specified, elements with weights are interleaved.  If a weight is 0, either
because it is the default weight or it was not specified, then elements are
delivered sequentially and all available data from one element is sent before
any data from the next.  If a given priority level has some elements with

> weights and some without, the two groups share the available bandwidth equally.

Reviewer 7: You can have 10 elements with weights at some priority level, sharing bandwidth.  When you add a single element to that priority level without a weight, the bandwidth for the 10 elements is halved.  This does not seem right.

```suggestion
weights and some without, the two groups share the available bandwidth proportionally to the number of elements in each group.
```

Reviewer 8: I suggest looking at rmarx proposal.

Author: Good point @igorlord This was reflecting what Patrick Meenan wrote up, but I agree that it's not clear it's the right option.

Personally, I would prefer removing weights from streams and only having one bit to declare that this resource is incrementally useful.

Like streams, placeholders have priority information associated with them.

When a placeholder is reparented, given a new weight or given a new strict

> priority, all dependent placeholders and streams are also re-prioritized.

Reviewer 4: ```suggestion
priority, the effective priority of all dependent placeholders and streams is impacted.
```
Just to be explicit that their priorities don't directly change.
